### ISSUE TITLE
Inopportune panic over LE rate-limiting

### ISSUE THREAD

**Issue by kseistrup** [mholt/caddy#642]
If Caddy receives a
```txt
[sub.example.com] failed to get certificate: acme: Error 429 - urn:acme:error:rateLimited - Error creating new cert :: Too many certificates already issued for: example.com
```
while attempting to renew a certiciate, it exits with an error — even if the certificate for said domain is still valid for ~3 weeks — rather than chilling and giving it a spin again tomorrow. As a consequence,  HTTPS has to be manually deactivated (and retried until it succeeds) for the domain.

As long as a certificate is still valid, Caddy shouldn't panic over being rate-limited by Let's Encrypt.

![keep-calm-and-notify-the-bofh](https://cloud.githubusercontent.com/assets/25586/13379021/7195e56a-de1a-11e5-9568-4fe1c64f6f24.png)

**Comment by kseistrup** [comment]
Oh, I forgot:
```sh
$ caddy -version
Caddy 0.8.2
```

**Comment by mholt** [comment]
This is not a panic. :wink:

This is actually expected behavior, since errors that occur while obtaining or renewing certificates when Caddy starts are considered fatal. This is because we cannot start the server on good faith if certificates failed to be obtained or renewed. If it failed now, what guarantee is there that it will not fail later?

While the server is running, we do retry as that is all we can do. But at start up, it would be disingenuous for the server to gloss over the error and just "hope it works later" while the user is still there.

I strongly prefer explicit error handling/reporting where possible, and this is one of those cases; especially with regards to security.

**Comment by kseistrup** [comment]
I beg to differ. It's not about “obtaining or renewing certificates”. I'm talking about using an already obtained certificate while it is still valid.

If there's a valid certificate, Caddy could use that one and notify the sysadmin. The current behaviour shuts down the entire service. By using the still valid certificate, everything could continue and the administrator is notified.

It could even be possible to introduce a softfail limit (say 1 month) and a hardfail limit (say 1 week): So if there's a working certificate and if we're between the softfail and the hardfail limit, use the “old” cetificate and notify the administrator. If we have passed the hardfail limit and the certicifate is still valid: ignore the service that makes use of said certificate and notify the administrator.

In my setup Caddy handles more than a couple of scores of domains and subdomains. With the current behavious Caddy shuts down serving all of the domains and subdomains, just because one domain with its subdomains are rate-limited. That's unacceptable.

**Comment by mholt** [comment]
> The current behaviour shuts down the entire service.

Technically, the service never gets started. Which is good, because something is broken. It's the only way to guarantee the operator knows that something is seriously wrong.

> It could even be possible to introduce a softfail limit (say 1 month) and a hardfail limit (say 1 week)

I am against adding complexity to this already-complex process. Especially when it comes to "ignoring the service" -- this is kind of against Caddy's philosophy.

> With the current behavious Caddy shuts down serving all of the domains and subdomains, just because one domain with its subdomains are rate-limited. That's unacceptable.

Well, none of them get started in the first place. That *is* acceptable when there is a blocking error that prevents Caddy from doing what you expect it to do. You expect it to keep your domains renewed, but there is clearly something wrong that is preventing that which the operator should fix before continuing.

**Comment by kseistrup** [comment]
Whether things even “get started in the first place” or are shut down is just a war on semantics, the end result is the same.

Following your logic, a mail server who has some greylisted and some non-greylisted emails in its outbound spool directory should refuse to start at all, thus disabling delivery of all emails, just because a few emails are greylisted.

My Caddy installation ran into all of this because I upgraded from v0.8 (that apparently didn't renew certificates) to v0.8.2 when there were less than a month left for some of the certificates. Had I kept v0.8 no certificates had been renewed at all. Now I have to manually enabled/disable HTTPS for one or more of the domains on a daily basis, hoping that eventually the webserver will start. And I even have to pray that more domains and subdomains are not added to the rate-limiting list, or this circus could go on for months.

Caddy's grave mistake is that it treats a temporary error as a permanent one and draws a wrong conclusion.  ALL of the certificates for ALL of the domains and subdomains are still valid for at least 3 weeks, yet you say that “there is a blocking error that prevents Caddy from doing what you expect it to do”. You're exactly wrong: there is no error except for a temporary refusal from LE's side with plenty of time to retry the renewal procedure tomorrow or the day after tomorrow. Yes, it may fail for the entire 3 weeks, but chances are that it will succeed (and run flawlessly until it does). But as long as Caddy deems that this is a fatal error that should be corrected before it is willing to start, it WILL faill 100% of the time — yes, it may even reduce the likelyhood for succeeding because new “expired“ domains could be added in the mean time.

Please allow Caddy to consider LE's rate-limiting the same way a mailserver would treat a greylisting or a temporary error, because such is the nature of this type of error.

I wish more people would chime in on this issue so that we could have it enlightened from all sides…

**Comment by jahmad** [comment]
It is a bit strange that Caddy introduced on-demand tls, and yet refused to start when valid certificates already available.

How about putting connecting to LE into conditional on startup, only do it if it's needed? will it be too complex?

**Comment by klingtnet** [comment]
I can't use my certificate backup to prevent certificate renewal on a fresh caddy installation and went into a rate limiting issue because of caddy `0.8.2`, see #647.

**Comment by mholt** [comment]
Give me a moment to think about this issue in some detail and model the problem space in my head. I'll sleep on it.

**Comment by jahmad** [comment]
Browsing issues brought me to this commit, 09a7af8cae. reading your comment there, that's the whole point of this request. 

>  we already have what we need on disk.

**Comment by mholt** [comment]
I've given this a lot of thought and discussed with other developers about their preferences based on their expertise. The reason I do not lightly change this behavior is because errors, especially while the operator is still present and the server is not yet started, are ripe to be handled by the user and be fixed immediately.

In this case, however, there is nothing the user can do to relieve the error except wait for time to pass. This is, as you've made abundantly clear, very frustrating with multiple sites which don't have problems.

Be assured I have read and re-read everything you have typed. I will address all of it. Hear me out and then I will tell you what I've decided to do about this, despite going slightly against my better judgment. Still, I hope it will bring a better experience for Caddy users.

@kseistrup 
> Following your logic, a mail server who has some greylisted and some non-greylisted emails in its outbound spool directory should refuse to start at all, thus disabling delivery of all emails, just because a few emails are greylisted.
> ...
> Please allow Caddy to consider LE's rate-limiting the same way a mailserver would treat a greylisting or a temporary error, because such is the nature of this type of error.

I disagree with the premise for this analogy. Mail is to a mail server as HTTP requests are to an HTTP server. Indeed Caddy should never be brought offline because of bad HTTP requests. But imagine that mail server trying to start up without proper credentials it needs to send the mail; it can only get those from the user, so it must refuse to start. This is somewhat akin to Caddy obtaining and renewing certificates. If a certificate cannot be obtained, the site owner has to try to fix the problem.

> I upgraded from v0.8 (that apparently didn't renew certificates) ... Had I kept v0.8 no certificates had been renewed at all.

Caddy 0.8 does in fact renew certificates. It renews them 14 days out, whereas version 0.8.2 renews them 30 days out. My guess is that your certificates were not in the old renewal window, but are in the new one. Perhaps similar reasoning to issue #648 in your case.

We made this change because Let's Encrypt sends renewal emails about 3-4 weeks out. We also realized that some rate limits that last a full week means that a 2 week window to renew certificates was cutting it close. Anyway, the 30 day renewal window is better.

> And I even have to pray that more domains and subdomains are not added to the rate-limiting list, or this circus could go on for months.

Let's Encrypt's rate limits are [documented](https://community.letsencrypt.org/t/rate-limits-for-lets-encrypt/6769) and so is the fact that Caddy uses Let's Encrypt. Given that these factors -- including the CA that it used -- are out of Caddy's control, it's the site owner's responsibility to stay within the CA's limits. As frustrating as this may be, I do affirm operator error as the cause of the rate limiting. (Unless a confirmed buggy loop is found in Caddy, which we haven't found one.)

As for praying that things will work: that is what you will have to do if Caddy continues to start up even though it knows there are errors. The advantage of the current behavior is that you are forced know that there's something wrong and fix it. No praying necessary. :wink:

> Caddy's grave mistake is that it treats a temporary error as a permanent one and draws a wrong conclusion.

Yes, you're right. There's even [a comment in the code](https://github.com/mholt/caddy/blob/741880a38be0104fb2023b944ea3b9afbc130e83/caddy/https/maintain.go#L93) to prove it. We assume the user wants the server to stop and report the error, rather than continuing with impending doom.

> ALL of the certificates for ALL of the domains and subdomains are still valid for at least 3 weeks, yet you say that “there is a blocking error that prevents Caddy from doing what you expect it to do”. You're exactly wrong: there is no error except for a temporary refusal from LE's side with plenty of time to retry the renewal procedure tomorrow or the day after tomorrow.

In this sense, you're right.

> Yes, it may fail for the entire 3 weeks, but chances are that it will succeed (and run flawlessly until it does).

We actually don't know for how long the error will block. We don't know if it will ever go away, even. Bug in the CA software? Other type of error (rate limits aren't in the ACME spec, and every CA will do them differently)? 

>  But as long as Caddy deems that this is a fatal error that should be corrected before it is willing to start, it WILL faill 100% of the time — yes, it may even reduce the likelyhood for succeeding because new “expired“ domains could be added in the mean time.

Yes, it will refuse to start 100% of the time, but I don't see that as a failure. The failure was what caused the error that prohibited Caddy from obtaining the certificates it needs.

"But it doesn't need it yet, only in 3 weeks."

Okay, that's true. So we let Caddy start. It runs for 3 weeks then maybe the site goes red.

What if the certificate expired in 3 days? Maybe enough time to get a successful renewal.

What about 3 hours? Ooo, cutting it close.

3 minutes? Caddy starts and 3 minutes later your site goes red. How nice of your web server to do that.

3 seconds? It's totally possible.

Point is, this rabbit hole is deep. I don't wanna go down it.

@jahmad 
> It is a bit strange that Caddy introduced on-demand tls, and yet refused to start when valid certificates already available.

That is kinda funny. :smile: 

> How about putting connecting to LE into conditional on startup, only do it if it's needed? will it be too complex?

"Needed" -- as in, the certificate is already expired and *must* be renewed before starting?

That goes back to the rabbit hole I just mentioned. Why wait until it's expired? Wouldn't you want to renew it if it expires in 3 seconds? Or 3 minutes or hours? Maybe 3 days would be fine, but not much closer.

You can always bring your own certificates and specify them in the Caddyfile. Expired ones, even. Those'll work. Your site will be red but I guess that's what you intended, since Caddy isn't managing the certificates you provide.

@klingtnet 
> I can't use my certificate backup to prevent certificate renewal on a fresh caddy installation and went into a rate limiting issue because of caddy 0.8.2

We figured this out over in your issue, I believe. It's due to the larger renewal window. Sorry about that; it's a one-time thing that only affects very few users and even fewer so adversely.

@jahmad 
> Browsing issues brought me to this commit, 09a7af8. reading your comment there, that's the whole point of this request.

That's a valid point.

Okay, phew. I think I've addressed all the concerns and cleared up misconceptions and thought about this enough to make a decision.

But I have to define semantics here in order for this to work. The crux of the issue is that Caddy "has it what it needs" at start up, so why refuse to start? The problem is that "what it needs" is ambiguous. Is a certificate that expires in 3 seconds "what it needs"? Is a certificate that expired 2 years ago "what it needs"? Is there even a difference?

These are hard questions. And they are the same rabbit hole I keep talking about.

I will define "what it needs" as **an unexpired certificate**. Whether that certificate expires in 3 weeks or in 3 seconds, Caddy has what it needs right then and that's apparently all that matters. It's semantics, it's annoying, but it will have to do. Building custom renewal timers just for certificates that are "too close" to expiring will not solve the problem since only time and/or the CA can lift rate limits.

So I will change [this if statement](https://github.com/mholt/caddy/blob/741880a38be0104fb2023b944ea3b9afbc130e83/caddy/https/maintain.go#L92) to read `if client.AllowPrompts && timeLeft < 0 {` which means, essentially: "If the server is still starting up, AND the certificate is expired." So Caddy will only refuse to start if the certificate it needs is expired and fails to renew. Otherwise it will log the error and continue to run.

To recap, comparing the current (old) behavior and the proposed (new) behavior:

**Current (refuse to start if renewal fails, always)**
- Advantage: Caddy doesn't mislead the operator by starting to serve a site when it knows there is a problem that will bring the site down now or in the future.
- Disadvantage: Caddy doesn't start, so no sites get served, even if other sites are healthy.

**Proposed (refuse to start only if renewal fails on expired certificates)**
- Advantage: Caddy will log any renewal errors and still start, serving all the healthy sites.
- Disadvantage: The site may go down in the near future and the user will only know if they watch the log output from startup; use has to "pray" that the error will work itself out.

**Comment by klingtnet** [comment]
@mholt Thank you for your detailed explanations!

> You can always bring your own certificates and specify them in the Caddyfile. Expired ones, even. Those'll work.

I didn't thought about that, this is a nice workaround for my problem. I can explicitly specify the certificates of my backup and use them until the rate limiting is gone or the new caddy version is released.

**Comment by xenolf** [comment]
@klingtnet Your server block should only have one `tls` keyword. It's `tls cert key` as per https://caddyserver.com/docs/tls

**Comment by klingtnet** [comment]
@xenolf I've totally overlooked that! How embarrassing :pensive:

**Comment by mholt** [comment]
Actually now you can have two or more `tls` directives as long as they don't conflict with each other (in other words, you can use more to provide more certificates but don't use one to say `tls off` while in another you turn it on).
